File: java/exercises/practice/unit_test_6dbfbc4d_2002_45bc_9b13_f0bf1c9f533e/distributed_tx/src/test/java/CoordinatorTest.java
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.*;

public class CoordinatorTest {

    // Mock service implementation for testing
    class MockService implements Service {
        private final String name;
        private boolean prepared = false;
        private boolean committed = false;
        private boolean rolledBack = false;
        private final long delay; // delay in milliseconds
        private final boolean failOnPrepare;
        private boolean busy = false; // simulate one transaction at a time

        public MockService(String name) {
            this(name, 0, false);
        }

        public MockService(String name, long delay, boolean failOnPrepare) {
            this.name = name;
            this.delay = delay;
            this.failOnPrepare = failOnPrepare;
        }

        @Override
        public synchronized boolean prepare(String transactionId) {
            // if already busy with another transaction, return false
            if (busy) {
                return false;
            }
            busy = true;
            try {
                if (delay > 0) {
                    Thread.sleep(delay);
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                busy = false;
                return false;
            }
            if (failOnPrepare) {
                busy = false;
                return false;
            }
            prepared = true;
            return true;
        }

        @Override
        public synchronized void commit(String transactionId) {
            if (!prepared) {
                return;
            }
            committed = true;
            busy = false;
        }

        @Override
        public synchronized void rollback(String transactionId) {
            rolledBack = true;
            busy = false;
        }

        @Override
        public String getServiceName() {
            return name;
        }

        public boolean isCommitted() {
            return committed;
        }

        public boolean isRolledBack() {
            return rolledBack;
        }
    }

    @Test
    public void testSuccessfulTransaction() throws Exception {
        // Create a coordinator with a 1 second timeout for prepare phase.
        Coordinator coordinator = new Coordinator(1000);
        String txId = coordinator.beginTransaction();

        // Create services that succeed in preparing.
        MockService inventory = new MockService("Inventory");
        MockService payment = new MockService("Payment");
        MockService order = new MockService("Order");

        coordinator.registerService(inventory, txId);
        coordinator.registerService(payment, txId);
        coordinator.registerService(order, txId);

        boolean prepared = coordinator.prepareTransaction(txId);
        assertTrue(prepared, "All services should prepare successfully.");
        coordinator.commitTransaction(txId);

        assertTrue(inventory.isCommitted(), "Inventory service should be committed.");
        assertTrue(payment.isCommitted(), "Payment service should be committed.");
        assertTrue(order.isCommitted(), "Order service should be committed.");
    }

    @Test
    public void testPrepareFailure() throws Exception {
        Coordinator coordinator = new Coordinator(1000);
        String txId = coordinator.beginTransaction();

        // Simulate Payment service failure by returning false on prepare.
        MockService inventory = new MockService("Inventory");
        MockService payment = new MockService("Payment", 0, true);
        MockService order = new MockService("Order");

        coordinator.registerService(inventory, txId);
        coordinator.registerService(payment, txId);
        coordinator.registerService(order, txId);

        boolean prepared = coordinator.prepareTransaction(txId);
        assertFalse(prepared, "Prepare should fail due to one service failing to prepare.");
        coordinator.rollbackTransaction(txId);

        assertTrue(inventory.isRolledBack(), "Inventory should be rolled back.");
        assertTrue(payment.isRolledBack(), "Payment should be rolled back.");
        assertTrue(order.isRolledBack(), "Order should be rolled back.");
    }

    @Test
    public void testTimeoutDuringPrepare() throws Exception {
        // Set coordinator timeout to 500ms while one service delays for 1000ms.
        Coordinator coordinator = new Coordinator(500);
        String txId = coordinator.beginTransaction();

        MockService inventory = new MockService("Inventory");
        // This service will delay, causing timeout.
        MockService payment = new MockService("Payment", 1000, false);
        MockService order = new MockService("Order");

        coordinator.registerService(inventory, txId);
        coordinator.registerService(payment, txId);
        coordinator.registerService(order, txId);

        boolean prepared = coordinator.prepareTransaction(txId);
        assertFalse(prepared, "Prepare should fail due to timeout in one service.");
        coordinator.rollbackTransaction(txId);

        assertTrue(inventory.isRolledBack(), "Inventory should be rolled back after timeout.");
        assertTrue(payment.isRolledBack(), "Payment should be rolled back after timeout.");
        assertTrue(order.isRolledBack(), "Order should be rolled back after timeout.");
    }

    @Test
    public void testConcurrentTransactions() throws Exception {
        final Coordinator coordinator = new Coordinator(1000);
        int numTransactions = 5;
        ExecutorService executor = Executors.newFixedThreadPool(numTransactions);
        List<Future<Boolean>> results = new ArrayList<>();

        for (int i = 0; i < numTransactions; i++) {
            results.add(executor.submit(() -> {
                String txId = coordinator.beginTransaction();
                MockService service1 = new MockService("Service1");
                MockService service2 = new MockService("Service2");

                coordinator.registerService(service1, txId);
                coordinator.registerService(service2, txId);

                boolean prepared = coordinator.prepareTransaction(txId);
                if (prepared) {
                    coordinator.commitTransaction(txId);
                    return service1.isCommitted() && service2.isCommitted();
                } else {
                    coordinator.rollbackTransaction(txId);
                    return service1.isRolledBack() && service2.isRolledBack();
                }
            }));
        }

        for (Future<Boolean> future : results) {
            try {
                assertTrue(future.get(2, TimeUnit.SECONDS), "Concurrent transaction failed.");
            } catch (Exception e) {
                fail("Concurrent transaction encountered an exception: " + e.getMessage());
            }
        }
        executor.shutdown();
    }
}

File: go/exercises/practice/unit_test_d6e2b458_0740_445d_abcf_6892ba7aff34/dist_tx_coord/coordinator_test.go
package dist_tx_coord

import (
	"errors"
	"testing"
	"time"
)

type mockService struct {
	name           string
	prepareErr     error
	commitErr      error
	prepareDelay   time.Duration
	commitDelay    time.Duration
	prepared       bool
	committed      bool
	rolledBack     bool
}

func (m *mockService) Prepare() error {
	time.Sleep(m.prepareDelay)
	m.prepared = true
	return m.prepareErr
}

func (m *mockService) CommitRollback(commit bool) error {
	time.Sleep(m.commitDelay)
	if commit {
		m.committed = true
		return m.commitErr
	}
	m.rolledBack = true
	return m.commitErr
}

func (m *mockService) GetName() string {
	return m.name
}

func TestTransactionCoordinator(t *testing.T) {
	t.Run("successful transaction", func(t *testing.T) {
		tc := NewTransactionCoordinator(10, 100)
		txID := tc.Begin()

		services := []Service{
			&mockService{name: "Inventory", prepareDelay: 50 * time.Millisecond},
			&mockService{name: "Payment", prepareDelay: 30 * time.Millisecond},
			&mockService{name: "Shipping", prepareDelay: 20 * time.Millisecond},
		}

		for _, svc := range services {
			if err := tc.Register(txID, svc); err != nil {
				t.Fatalf("Register failed: %v", err)
			}
		}

		if err := tc.End(txID, true); err != nil {
			t.Errorf("End failed: %v", err)
		}

		for _, svc := range services {
			mock := svc.(*mockService)
			if !mock.prepared {
				t.Errorf("%s was not prepared", mock.name)
			}
			if !mock.committed {
				t.Errorf("%s was not committed", mock.name)
			}
		}
	})

	t.Run("failed prepare phase", func(t *testing.T) {
		tc := NewTransactionCoordinator(10, 100)
		txID := tc.Begin()

		services := []Service{
			&mockService{name: "Inventory", prepareDelay: 50 * time.Millisecond},
			&mockService{name: "Payment", prepareDelay: 30 * time.Millisecond, prepareErr: errors.New("insufficient funds")},
			&mockService{name: "Shipping", prepareDelay: 20 * time.Millisecond},
		}

		for _, svc := range services {
			if err := tc.Register(txID, svc); err != nil {
				t.Fatalf("Register failed: %v", err)
			}
		}

		err := tc.End(txID, true)
		if err == nil {
			t.Error("Expected error but got none")
		}

		for _, svc := range services {
			mock := svc.(*mockService)
			if mock.name != "Payment" && !mock.prepared {
				t.Errorf("%s was not prepared", mock.name)
			}
			if mock.committed {
				t.Errorf("%s was committed but shouldn't be", mock.name)
			}
			if mock.name != "Payment" && !mock.rolledBack {
				t.Errorf("%s was not rolled back", mock.name)
			}
		}
	})

	t.Run("timeout during prepare", func(t *testing.T) {
		tc := NewTransactionCoordinator(10, 100)
		txID := tc.Begin()

		services := []Service{
			&mockService{name: "Inventory", prepareDelay: 600 * time.Millisecond},
			&mockService{name: "Payment", prepareDelay: 30 * time.Millisecond},
			&mockService{name: "Shipping", prepareDelay: 20 * time.Millisecond},
		}

		for _, svc := range services {
			if err := tc.Register(txID, svc); err != nil {
				t.Fatalf("Register failed: %v", err)
			}
		}

		err := tc.End(txID, true)
		if err == nil {
			t.Error("Expected timeout error but got none")
		}

		for _, svc := range services {
			mock := svc.(*mockService)
			if mock.name == "Inventory" && mock.prepared {
				t.Errorf("Inventory should have timed out before preparing")
			}
			if mock.committed {
				t.Errorf("%s was committed but shouldn't be", mock.name)
			}
		}
	})

	t.Run("concurrent transactions", func(t *testing.T) {
		tc := NewTransactionCoordinator(10, 100)
		var txIDs []TransactionID

		// Start 5 concurrent transactions
		for i := 0; i < 5; i++ {
			txID := tc.Begin()
			txIDs = append(txIDs, txID)

			services := []Service{
				&mockService{name: "Inventory", prepareDelay: time.Duration(i*10) * time.Millisecond},
				&mockService{name: "Payment", prepareDelay: time.Duration(i*5) * time.Millisecond},
			}

			for _, svc := range services {
				if err := tc.Register(txID, svc); err != nil {
					t.Fatalf("Register failed: %v", err)
				}
			}
		}

		// Commit all transactions
		for _, txID := range txIDs {
			if err := tc.End(txID, true); err != nil {
				t.Errorf("End failed for tx %v: %v", txID, err)
			}
		}
	})

	t.Run("register after end", func(t *testing.T) {
		tc := NewTransactionCoordinator(10, 100)
		txID := tc.Begin()

		svc := &mockService{name: "Inventory"}
		if err := tc.Register(txID, svc); err != nil {
			t.Fatalf("Initial register failed: %v", err)
		}

		if err := tc.End(txID, true); err != nil {
			t.Fatalf("End failed: %v", err)
		}

		err := tc.Register(txID, &mockService{name: "LateService"})
		if err == nil {
			t.Error("Expected error when registering after end but got none")
		}
	})
}